File: patterns/08_strategy/php/ars_bellica.php
<?php
/**
 * STRATEGY PATTERN — Roman Battle Formations
 * Run: php ars_bellica.php
 */
declare(strict_types=1);

interface BattleStrategy {
    public function strategyName(): string;
    public function execute(array $s): string;
}
class Testudo implements BattleStrategy {
    public function strategyName(): string { return "TESTUDO (Tortoise)"; }
    public function execute(array $s): string {
        return "🐢 Shields locked! vs {$s['enemies']} on {$s['terrain']}. Casualties -60%!";
    }
}
class Cuneus implements BattleStrategy {
    public function strategyName(): string { return "CUNEUS (Wedge)"; }
    public function execute(array $s): string {
        return "🔺 CHARGE! Wedge pierces {$s['enemies']} on {$s['terrain']}!";
    }
}
class Orbis implements BattleStrategy {
    public function strategyName(): string { return "ORBIS (Circle)"; }
    public function execute(array $s): string {
        return "⭕ 360° defence! Morale {$s['morale']}/10. Holding!";
    }
}
class RomanLegion {
    public function __construct(private string $name, private BattleStrategy $strategy) {}
    public function setStrategy(BattleStrategy $s): void {
        echo "  ⚙  {$this->name}: {$this->strategy->strategyName()} → {$s->strategyName()}\n";
        $this->strategy = $s;
    }
    public function engage(array $s): void {
        echo "\n  🦅 {$this->name} [{$this->strategy->strategyName()}]\n";
        echo "     " . $this->strategy->execute($s) . "\n";
    }
}

echo "╔═══════════════════════════════════════════════╗\n";
echo "║   STRATEGY PATTERN — PHP                      ║\n";
echo "║   Roman Battle Formations                     ║\n";
echo "╚═══════════════════════════════════════════════╝\n\n";
$legion = new RomanLegion("Legio X Gemina", new Testudo());
$legion->engage(['enemies'=>500, 'terrain'=>'open field', 'morale'=>8]);
echo "\n── SCIPIO ADAPTS ────────────────────────────────\n";
$legion->setStrategy(new Cuneus());
$legion->engage(['enemies'=>200, 'terrain'=>'open field', 'morale'=>9]);
$legion->setStrategy(new Orbis());
$legion->engage(['enemies'=>5000, 'terrain'=>'valley', 'morale'=>4]);
echo "\n\"Non una via vincimus!\"\n";

File: patterns/08_strategy/python/ars_bellica.py
#!/usr/bin/env python3
"""
STRATEGY PATTERN — Roman Battle Formations
"Non una via vincimus" — We do not win by one path
Run: python3 ars_bellica.py
"""
from abc import ABC, abstractmethod
from dataclasses import dataclass

@dataclass
class BattleSituation:
    enemies: int
    terrain: str
    morale:  int

class BattleStrategy(ABC):
    @abstractmethod
    def strategy_name(self) -> str: ...
    @abstractmethod
    def execute(self, s: BattleSituation) -> str: ...

class Testudo(BattleStrategy):
    def strategy_name(self) -> str: return "TESTUDO (Tortoise)"
    def execute(self, s: BattleSituation) -> str:
        return f"🐢 Shields locked! vs {s.enemies} on {s.terrain}. Casualties -60%!"

class Cuneus(BattleStrategy):
    def strategy_name(self) -> str: return "CUNEUS (Wedge)"
    def execute(self, s: BattleSituation) -> str:
        return f"🔺 CHARGE! Wedge pierces {s.enemies} on {s.terrain}!"

class Orbis(BattleStrategy):
    def strategy_name(self) -> str: return "ORBIS (Circle)"
    def execute(self, s: BattleSituation) -> str:
        return f"⭕ 360° defence! Morale {s.morale}/10. Holding!"

class RomanLegion:
    def __init__(self, name: str, strategy: BattleStrategy):
        self._name     = name
        self._strategy = strategy
    def set_strategy(self, strategy: BattleStrategy):
        print(f"  ⚙  {self._name}: {self._strategy.strategy_name()} → {strategy.strategy_name()}")
        self._strategy = strategy
    def engage(self, s: BattleSituation):
        print(f"\n  🦅 {self._name} [{self._strategy.strategy_name()}]")
        print(f"     {self._strategy.execute(s)}")

def main():
    print("╔═══════════════════════════════════════════════╗")
    print("║   STRATEGY PATTERN — Python                   ║")
    print("║   Roman Battle Formations                     ║")
    print("╚═══════════════════════════════════════════════╝\n")
    legion = RomanLegion("Legio X Gemina", Testudo())
    legion.engage(BattleSituation(500, "open field", 8))
    print("\n── SCIPIO ADAPTS TO THE BATTLEFIELD ────────────")
    legion.set_strategy(Cuneus())
    legion.engage(BattleSituation(200, "open field", 9))
    legion.set_strategy(Orbis())
    legion.engage(BattleSituation(5000, "surrounded valley", 4))
    print('\n"Non una via vincimus!"')

if __name__ == "__main__":
    main()
